include/seastar/core/shared_mutex.hh

```diff
@@ -234,6 +234,122 @@ with_lock(shared_mutex& sm, Func&& func) noexcept {
     }
 }
 
+/// \brief RAII-based shared lock for a \ref shared_mutex object.
+///
+/// As soon as the destructor of an object of this class is called,
+/// the corresponding \ref shared_mutex is shared unlocked.
+///
+/// The class is effectively a safe way to keep a \ref shared_mutex
+/// locked and to unconditionally release it because no matter what
+/// happens, the destructor should eventually be called.
+///
+/// The corresponding \ref shared_mutex object MUST be kept alive
+/// throughout the lifetime of this object.
+///
+/// The class is NOT responsible for shared locking a \ref shared_mutex.
+/// \see get_shared_lock(shared_mutex&)
+/// \see unique_lock
+class shared_lock {
+private:
+    friend future<shared_lock> get_shared_lock(shared_mutex& mutex);
+private:
+    shared_mutex* _mutex;
+
+private:
+    shared_lock(shared_mutex& mutex) noexcept : _mutex{std::addressof(mutex)} {}
+public:
+    shared_lock(shared_lock&& other) noexcept : _mutex{std::exchange(other._mutex, nullptr)} {}
+    shared_lock& operator=(shared_lock&& other) noexcept {
+        if (this != std::addressof(other)) {
+            this->~shared_lock();
+            new (this) shared_lock{std::move(other)};
+        }
+        return *this;
+    }
+
+    shared_lock(const shared_lock&) = delete;
+    shared_lock& operator=(const shared_lock&) = delete;
+
+    ~shared_lock() noexcept {
+        if (_mutex) {
+            _mutex->unlock_shared();
+            _mutex = nullptr;
+        }
+    }
+};
+
+/// \brief Construct a RAII-based shared lock corresponding to a given \ref shared_mutex object.
+///
+/// Since constructors cannot be exectued asynchronously, use this function
+/// to shared lock a \ref shared_mutex and construct a \ref shared_lock object
+/// that will unlock the mutex when it's being destroyed.
+///
+/// The caller is responsible for keeping the corresponding \ref shared_mutex object
+/// alive at least until the returned \ref shared_lock is destroyed.
+inline future<shared_lock> get_shared_lock(shared_mutex& mutex) {
+    return mutex.lock_shared().then([&mutex] {
+        return make_ready_future<shared_lock>(shared_lock{mutex});
+    });
+}
+
+/// \brief RAII-based unique lock for a \ref shared_mutex object.
+///
+/// As soon as the destructor of an object of this class is called,
+/// the corresponding \ref shared_mutex is unlocked.
+///
+/// The class is effectively a safe way to keep a \ref shared_mutex
+/// locked and to unconditionally release it because no matter what
+/// happens, the destructor should eventually be called.
+///
+/// The corresponding \ref shared_mutex object MUST be kept alive
+/// throughout the lifetime of this object.
+///
+/// The class is NOT responsible for locking a \ref shared_mutex.
+/// \see get_unique_lock(shared_mutex&)
+/// \see shared_lock
+class unique_lock {
```

Comment:
The counterpart of shared lock is typically an exclusive lock, not unique.

Let's stay consistent with:
```
    /// Lock the \c shared_mutex for exclusive access
    ///
    /// \return a future that becomes ready when no access, shared or exclusive
    ///         is granted to anyone.
    future<> lock() noexcept {
```

Reply:
The C++ standard library has std::shared_lock and std::unique_lock, and I assume this is wheret his choice of names comes from. The questions I raised in my review was why can't we use std::shared_lock and std::unique_lock themselves, and not just their names.